crates/autopilot/src/decoded_settlement.rs

```diff
@@ -197,16 +175,68 @@ impl TryFrom<database::orders::OrderExecution> for OrderExecution {
 }
 
 impl DecodedSettlement {
+    /// Number of bytes that may be appended to the calldata to store an auction
+    /// id.
+    pub const META_DATA_LEN: usize = 8;
+
     pub fn new(input: &[u8]) -> Result<Self, DecodingError> {
         let function = GPv2Settlement::raw_contract()
             .abi
             .function("settle")
             .unwrap();
-        let decoded_input = decode_function_input(function, input)?;
-        <DecodedSettlementTokenized>::from_token(Token::Tuple(decoded_input))
-            .map(Into::into)
-            .context("failed to decode settlement")
-            .map_err(From::from)
+        let without_selector = input
+            .strip_prefix(&function.selector())
+            .ok_or(DecodingError::InvalidSelector)?;
+
+        // Decoding calldata without expecting metadata can succeed even if metadata
+        // was appended. The other way around would not work so we do that first.
+        if let Ok(decoded) = Self::try_new(without_selector, function, true) {
+            return Ok(decoded);
+        }
+        Self::try_new(without_selector, function, false).map_err(Into::into)
+    }
+
+    fn try_new(data: &[u8], function: &Function, with_metadata: bool) -> Result<Self> {
+        let metadata_len = if with_metadata {
+            anyhow::ensure!(
+                data.len() % 32 == Self::META_DATA_LEN,
+                "calldata does not contain the expected number of bytes to include metadata"
+            );
```

Comment:
Apparently the calldata format is not guaranteed to have unique layout for a given ABI. However, in practice the ecosystem expects a certain ABI (it's already encoded in the `ethabi` crate so we wouldn't even be able to parse non-conforming calldata with or without appended `auction_id`). That's why I doubled down on this format and expect `calldata.len() % 32 == 12` when we try to extract the `auction_id`. 32 bytes for the word size (`uint256`), 4 bytes for the function selector, 8 bytes for the appended `auction_id`.
This function expects the function select to be removed already, hence `== Self::META_DATA_LEN`.